### PR TITLE
mon/OSDMonitor: list osd tree in named bucket

### PR DESCRIPTION
**Pull request by kungf**
easy to get osd status in named bucket.
eg.
"ceph osd tree-ls node-1" will only show osds in bucket node-1,

Signed-off-by: kungf <yang.wang@easystack.cn>
@yangdongsheng

### CODE REVIEW

File: src/crush/CrushTreeDumper.h [hunk 1]
```diff
@@ -167,6 +167,14 @@ namespace CrushTreeDumper {
...
     bool is_touched(int id) const { return touched.count(id) > 0; }
 
+    void set_root(const string& bucket) {
+      roots.clear();
+      if (crush->name_exists(bucket)) {
+	int i = crush->get_item_id(bucket);
+	roots.insert(i);
+      }
+    }
+
```
Comment (hjwsm1989) [inline src/crush/CrushTreeDumper.h]: roots is empty if the bucket doesn't exist?

Comment (kungf) [reply src/crush/CrushTreeDumper.h]: yes, nothing to show

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -3497,6 +3497,7 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
   }
   else if (prefix == "osd dump" ||
 	   prefix == "osd tree" ||
+	   prefix == "osd tree-ls" ||
 	   prefix == "osd ls" ||
```
Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: Seems by accident you changed the mode of this file to be `executable `...

Comment (kungf) [reply src/mon/OSDMonitor.cc]: thanks, i change it to default

**Comment by kungf** [comment]
@liewegas need a review, please

**Comment by liewegas** [comment]
Don't we already have 'ceph osd ls-tree <name>' that does exactly this?  (There is also 'ceph osd crush ls <name>' that returns the immediate child items.)

**Comment by kungf** [comment]
@liewegas ceph osd ls-tree only return osd id not with their status and weight, and ceph osd crush ls only return the the direct child items(children may also be bucket ). they don't show the osd status(up, down).  this pr cmd can list:
1  all osds with their status, 
2. can only show wanted status osds(like ceph osd tree up/down/destory),
3. show all osds in one bucket, the bucket may be host, rack, row, root ... so it's easy to get the osds in host/rack/row...

**Comment by liewegas** [comment]
what about s/tree-ls/tree-under/ or tree-from?  (The ls-tree vs tree-ls bugs me, and in the other cases 'ls' is always a raw list, not a table.)

**Comment by kungf** [comment]
@liewegas it's also hard for me to choose a proper command name .
what about tree-in?  tree-under is a little long,  tree-in may be easier to knock the keyboard

**Comment by kungf** [comment]
@liewegas cmd name was changed to tree-under, need a review, please

**Comment by yuriw** [comment]
[wip-yuri4-testing-2018-01-02-1711](https://trello.com/c/NEW8iWlF/216-wip-yuri4-testing-2018-01-02-1711)

**Pull request merged**